src/Security/Authentication/Certificate/samples/Certificate.Optional.Sample/Program.cs

```diff
@@ -22,23 +24,38 @@ public static void Main(string[] args)
         Host.CreateDefaultBuilder(args)
             .ConfigureWebHostDefaults(webBuilder =>
             {
+                // the certificate is issued for example.com and cert.example.com domains
+                var serverCertificate = CertificateLoader.LoadFromStoreCert(
+                    "example.com", "My", StoreLocation.CurrentUser,
+                    allowInvalid: true);
+
                 webBuilder.UseStartup<Startup>();
                 webBuilder.ConfigureKestrel((context, options) =>
                 {
-                        // Kestrel can't have different ssl settings for different hosts on the same IP because there's no way to change them based on SNI.
-                        // https://github.com/dotnet/runtime/issues/31097
-                        options.Listen(IPAddress.Parse(HostWithoutCert), 5001, listenOptions =>
-                    {
-                        listenOptions.UseHttps(httpsOptions =>
-                        {
-                            httpsOptions.ClientCertificateMode = ClientCertificateMode.NoCertificate;
-                        });
-                    });
-                    options.Listen(IPAddress.Parse(HostWithCert), 5001, listenOptions =>
+                    options.ListenAnyIP(5001, listenOptions =>
                     {
-                        listenOptions.UseHttps(httpsOptions =>
+                        listenOptions.UseHttps(new TlsHandshakeCallbackOptions()
```

Comment:
Important: see the old code and the issue https://github.com/dotnet/runtime/issues/31097
with the new API looks unnecessary to have the two calls to Listen because now every connection will make a call to `OnConnection ` instead of initializing a "state" for every connection; but I'm not 100% sure about this

Reply:
You're correct, only one Listen is needed here.